extension/src/paymentHandler/payment-handler.js

```diff
@@ -1,57 +1,42 @@
-const paypalHandler = require('../paymentHandler/paypal/paypal.handler')
-const kcpHandler = require('../paymentHandler/kcp/kcp-payment.handler')
-const refundHandler = require('../paymentHandler/cancel-or-refund.handler')
-const fetchPaymentMethodsHandler = require('../paymentHandler/fetch-payment-methods.handler')
-
-const paymentHandlers = {
-  creditCardHandler,
-  fetchPaymentMethodsHandler,
-  paypalHandler,
-  kcpHandler,
-  refundHandler
-}
+const getPaymentMethodsHandler = require('./get-payment-methods.handler')
+const getOriginKeysHandler = require('./get-origin-keys.handler')
 
 async function handlePayment (paymentObject) {
   const validatorBuilder = ValidatorBuilder.withPayment(paymentObject)
-  const adyenValidator = validatorBuilder
-    .validateAdyen()
+  const adyenValidator = validatorBuilder.validateAdyen()
   if (adyenValidator.hasErrors())
     // if it's not adyen payment, ignore the payment
     return { success: true, data: null }
-  const merchantReferenceValidator = validatorBuilder
-    .validateMerchantReferenceField()
-  if (merchantReferenceValidator.hasErrors())
-    return {
-      success: false,
-      data: merchantReferenceValidator.buildCtpErrorResponse()
-    }
-  const paymentMethodValidator = validatorBuilder.validatePaymentMethod()
-  if (paymentMethodValidator.hasErrors())
+
+  const requestValidator = validatorBuilder.validateRequestFields()
+  if (requestValidator.hasErrors())
     return {
       success: false,
-      data: paymentMethodValidator.buildCtpErrorResponse()
+      data: requestValidator.buildCtpErrorResponse()
     }
-  const handler = _getPaymentHandler(paymentObject)
-  const handlerResponse = await handler.handlePayment(paymentObject)
-  if (handlerResponse.errors)
-    return { success: false, data: handlerResponse }
+
+  const handlers = _getPaymentHandlers(paymentObject)
+  const handlerResponses = await Promise.all(
+    handlers.map(handler => handler.execute(paymentObject))
+  )
+  const handlerResponse = {
+    actions: handlerResponses.flatMap(result => result.actions)
+  }
   return { success: true, data: handlerResponse }
 }
 
-function _getPaymentHandler (paymentObject) {
-  const paymentValidator = ValidatorBuilder.withPayment(paymentObject)
-  if (paymentValidator.isCancelOrRefund())
-    return paymentHandlers.refundHandler
-  if (paymentValidator.isPaypal())
-    return paymentHandlers.paypalHandler
-  if (paymentValidator.isCreditCard())
-    return paymentHandlers.creditCardHandler
-  if (paymentValidator.isKcp())
-    return paymentHandlers.kcpHandler
-  return paymentHandlers.fetchPaymentMethodsHandler
-}
+function _getPaymentHandlers (paymentObject) {
+  // custom field on payment is not a mandatory field.
+  if (!paymentObject.custom)
+    return []
 
+  const handlers = []
+  if (paymentObject.custom.fields.getOriginKeysRequest && !paymentObject.custom.fields.getOriginKeysResponse)
```

Comment:
I guess we will have soon more of these if clauses. If you want, you can put these condition checkings in the handlers, put the handlers in an array and here just iterate through this array and call the checking methods. Sth like this:
```
const availableHandlers = [getPaymentMethodsHandler, getOriginKeysHandler]
const handlers = []
for (const handler of handlers)
    if (handler.isSuitable(paymentObject))
        handlers.push(handler)
return handlers
```
But maybe this is not so readable. You can decide.

Reply:
the structure looks ok, deciding which handler will work is not depend only on the handler, because handlers have side effects, let's say checking makePayment and submitPaymentDetails should not be work parallel together, so in makePayment handler checking submitPaymentDetail request will not be a good practice, probably validation should be outside the handler.

Reply:
OK let's wait and refactor later then.